From 719f82f28da148f6634be6be7bffd6a1b508e497 Mon Sep 17 00:00:00 2001
From: Chaozhong Yang <yangchaozhong.2009@gmail.com>
Date: Mon, 13 Feb 2017 16:16:23 -0800
Subject: [PATCH 0983/1164] CDH-50329: Backport HIVE-15872 : The
 PERCENTILE_APPROX UDAF does not work with empty
 set (Chaozhong Yang, reviewed by Wei Zheng)

Change-Id: I88292c41b71f58b597667051880020886e2217cd
---
 .../udf/generic/GenericUDAFPercentileApprox.java   |    2 ++
 .../clientpositive/udaf_percentile_approx_23.q     |    3 +++
 .../clientpositive/udaf_percentile_approx_23.q.out |   11 +++++++++++
 3 files changed, 16 insertions(+)

diff --git a/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFPercentileApprox.java b/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFPercentileApprox.java
index 56e76be..8692436 100644
--- a/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFPercentileApprox.java
+++ b/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFPercentileApprox.java
@@ -310,6 +310,8 @@ public void merge(AggregationBuffer agg, Object partial) throws HiveException {
           myagg.quantiles[i-1] = partialHistogram.get(i).get();
         }
         partialHistogram.subList(0, nquantiles+1).clear();
+      } else {
+        partialHistogram.subList(0, 1).clear();
       }
 
       // merge histograms
diff --git a/ql/src/test/queries/clientpositive/udaf_percentile_approx_23.q b/ql/src/test/queries/clientpositive/udaf_percentile_approx_23.q
index 028acdb..adf84b0 100644
--- a/ql/src/test/queries/clientpositive/udaf_percentile_approx_23.q
+++ b/ql/src/test/queries/clientpositive/udaf_percentile_approx_23.q
@@ -96,3 +96,6 @@ select percentile_approx(case when key < 100 then cast('NaN' as double) else key
 explain
 select percentile_approx(key, 0.5) from bucket;
 select percentile_approx(key, 0.5) between 255.0 and 257.0 from bucket;
+
+-- test where number of elements is zero
+select percentile_approx(key, array(0.50, 0.70, 0.90, 0.95, 0.99)) from bucket where key > 10000;
diff --git a/ql/src/test/results/clientpositive/udaf_percentile_approx_23.q.out b/ql/src/test/results/clientpositive/udaf_percentile_approx_23.q.out
index 14a6874..36d692a 100644
--- a/ql/src/test/results/clientpositive/udaf_percentile_approx_23.q.out
+++ b/ql/src/test/results/clientpositive/udaf_percentile_approx_23.q.out
@@ -626,3 +626,14 @@ POSTHOOK: type: QUERY
 POSTHOOK: Input: default@bucket
 #### A masked pattern was here ####
 true
+PREHOOK: query: -- test where number of elements is zero
+select percentile_approx(key, array(0.50, 0.70, 0.90, 0.95, 0.99)) from bucket where key > 10000
+PREHOOK: type: QUERY
+PREHOOK: Input: default@bucket
+#### A masked pattern was here ####
+POSTHOOK: query: -- test where number of elements is zero
+select percentile_approx(key, array(0.50, 0.70, 0.90, 0.95, 0.99)) from bucket where key > 10000
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@bucket
+#### A masked pattern was here ####
+NULL
-- 
1.7.9.5

